keep/rulesengine/rulesengine.py

```diff
@@ -0,0 +1,127 @@
+                self.logger.info("Rule applies, creating grouped alert")
+                # create grouped alert
+                event_payload = []
+                fingerprints = []
+                for group in rule_results:
+                    # todo: we take here the first but we should take all of them down the road
+                    group_payload = rule_results[group][0].dict().get("event")
+                    event_payload.append(group_payload)
+                    fingerprints.append(rule_results[group][0].fingerprint)
+                # TODO: should be calculated somehow else
+                fingerprint = hashlib.sha256(
+                    "".join(fingerprints).encode("utf-8")
+                ).hexdigest()
+                group_alert_name = f"Group alert {rule.name}: " + ", ".join(
+                    [event["name"] for event in event_payload]
+                )
+                create_alert_db(
+                    tenant_id=self.tenant_id,
+                    provider_type="rules",
+                    provider_id=rule.id,
+                    # todo: event should support list?
+                    event={
+                        "events": event_payload,
+                        "name": group_alert_name,
+                        "lastReceived": max(
+                            [event["lastReceived"] for event in event_payload]
+                        ),
+                        "severity": max([event["severity"] for event in event_payload]),
+                        "source": list(
+                            set([event["source"][0] for event in event_payload])
+                        ),
+                        # TODO: should be calculated somehow else
+                        "id": fingerprint,
+                        "status": "firing",
+                    },
+                    fingerprint=fingerprint,
+                )
+                self.logger.info("Created alert")
+
+        self.logger.info("Rules ran, alerts created")
+        return
+
+    def _extract_subrules(self, expression):
+        # CEL rules looks like '(source == "sentry") && (source == "grafana" && severity == "critical")'
+        # and we need to extract the subrules
+        sub_rules = expression.split(") && (")
+        # the first and the last rules will have a ( or ) at the beginning or the end
+        sub_rules[0] = sub_rules[0][1:]
+        sub_rules[-1] = sub_rules[-1][:-1]
+        return sub_rules
+
+    # TODO: a lot of unit tests to write here
+    def _run_rule_on_event(self, rule, event: AlertDto):
+        sub_rules = self._extract_subrules(rule.definition_cel)
+        payload = event.dict()
+        # workaround since source is a list
+        # todo: fix this in the future
+        payload["source"] = payload["source"][0]
+        for sub_rule in sub_rules:
+            env = celpy.Environment()
+            ast = env.compile(sub_rule)
+            prgm = env.program(ast)
+            r = prgm.evaluate(payload)
```

Comment:
some docs for this?

Reply:
isn't this "dangerous" (couldn't sub_rule have some malicious code?)

Reply:
Nope since its just Common Expression Logic (CEL) and its not actually running any code - https://github.com/google/cel-spec and the pyhton implementation - https://github.com/cloud-custodian/cel-python

anyway I've added some comments